File: Python/quicksort/Gemini/Gemini.py
from random import randint

def quicksort(arr):
  """Sorts an array of elements using the Quicksort algorithm."""

  if len(arr) <= 1:
    return arr

  # Choose a pivot element (you can use different strategies)
  pivot = arr[0]

  # Partition the array into two subarrays: elements smaller than pivot and elements greater than pivot
  left, right = [], []
  for num in arr[1:]:
    if num < pivot:
      left.append(num)
    else:
      right.append(num)

  # Recursively sort the subarrays and combine them with the pivot
  return quicksort(left) + [pivot] + quicksort(right)

def main():
    # Generate 1000 random values
    random_values = [randint(1, 10000) for _ in range(1000)]

    # Sort the generated random values using quicksort
    sorted_values = quicksort(random_values)

if __name__ == "__main__":
    main()

File: com/Java/quicksort/gemini/Gemini.java
public class Gemini {

    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i+1] and arr[high]
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }
  
    public static void main(String[] args) {
        Random random = new Random();
        int[] array = new int[1000];
        
        // Generate 1000 random values
        for (int i = 0; i < 1000; i++) {
            array[i] = random.nextInt(1000); // Random value between 0 and 999
        }

        // Sort the array
        quickSort(array, 0, array.length - 1);
    }

}